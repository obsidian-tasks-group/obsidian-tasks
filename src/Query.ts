import chrono from 'chrono-node';
import { LayoutOptions } from './LayoutOptions';

import { Status, Task } from './Task';

type Sorting = 'status' | 'due' | 'done' | 'path' | 'description';

export class Query {
    private _limit: number | undefined = undefined;
    private _layoutOptions: LayoutOptions = new LayoutOptions();
    private _filters: ((task: Task) => boolean)[] = [];
    private _error: string | undefined = undefined;
    private _sorting: Sorting[] = [];

    private readonly noDueString = 'no due date';
    private readonly dueRegexp = /^due (before|after|on)? ?(.*)/;

    private readonly doneString = 'done';
    private readonly notDoneString = 'not done';
    private readonly doneRegexp = /^done (before|after|on)? ?(.*)/;

    private readonly pathRegexp = /^path (includes|does not include) (.*)/;
    private readonly descriptionRegexp =
        /^description (includes|does not include) (.*)/;
    private readonly sortByRegexp =
        /^sort by (status|due|done|path|description)/;

    private readonly headingRegexp =
        /^heading (includes|does not include) (.*)/;

    private readonly hideOptionsRegexp =
        /^hide (task count|backlink|done date|due date|recurrence rule|edit button)/;

    private readonly recurringString = 'is recurring';
    private readonly notRecurringString = 'is not recurring';

    private readonly limitRegexp = /^limit (to )?(\d+)( tasks?)?/;
    private readonly excludeSubItemsString = 'exclude sub-items';

    constructor({ source }: { source: string }) {
        source
            .split('\n')
            .map((line: string) => line.trim())
            .forEach((line: string) => {
                switch (true) {
                    case line === '':
                        break;
                    case line === this.doneString:
                        this._filters.push(
                            (task) => task.status === Status.Done,
                        );
                        break;
                    case line === this.notDoneString:
                        this._filters.push(
                            (task) => task.status !== Status.Done,
                        );
                        break;
                    case line === this.recurringString:
                        this._filters.push(
                            (task) => task.recurrenceRule !== null,
                        );
                        break;
                    case line === this.notRecurringString:
                        this._filters.push(
                            (task) => task.recurrenceRule === null,
                        );
                        break;
                    case line === this.excludeSubItemsString:
                        this._filters.push((task) => task.indentation === '');
                        break;
                    case line === this.noDueString:
                        this._filters.push((task) => task.dueDate === null);
                        break;
                    case this.dueRegexp.test(line):
                        this.parseDueFilter({ line });
                        break;
                    case this.doneRegexp.test(line):
                        this.parseDoneFilter({ line });
                        break;
                    case this.pathRegexp.test(line):
                        this.parsePathFilter({ line });
                        break;
                    case this.descriptionRegexp.test(line):
                        this.parseDescriptionFilter({ line });
                        break;
                    case this.headingRegexp.test(line):
                        this.parseHeadingFilter({ line });
                        break;
                    case this.limitRegexp.test(line):
                        this.parseLimit({ line });
                        break;
                    case this.sortByRegexp.test(line):
                        this.parseSortBy({ line });
                        break;
                    case this.hideOptionsRegexp.test(line):
                        this.parseHideOptions({ line });
                        break;
                    default:
                        this._error = 'do not understand query';
                }
            });
    }

    public get limit(): number | undefined {
        return this._limit;
    }

    public get layoutOptions(): LayoutOptions {
        return this._layoutOptions;
    }

    public get filters(): ((task: Task) => boolean)[] {
        return this._filters;
    }

    public get sorting(): Sorting[] {
        return this._sorting;
    }

    public get error(): string | undefined {
        return this._error;
    }

    private parseHideOptions({ line }: { line: string }): void {
        const hideOptionsMatch = line.match(this.hideOptionsRegexp);
        if (hideOptionsMatch !== null) {
            const option = hideOptionsMatch[1].trim().toLowerCase();

            if (option === 'task count') {
                this._layoutOptions.hideTaskCount = true;
            } else if (option === 'backlink') {
                this._layoutOptions.hideBacklinks = true;
            } else if (option === 'done date') {
                this._layoutOptions.hideDoneDate = true;
            } else if (option === 'due date') {
                this._layoutOptions.hideDueDate = true;
            } else if (option === 'recurrence rule') {
                this._layoutOptions.hideRecurrenceRule = true;
            } else if (option === 'edit button') {
                this._layoutOptions.hideEditButton = true;
            } else {
                this._error = 'do not understand hide option';
            }
        }
    }

    private parseDueFilter({ line }: { line: string }): void {
        const dueMatch = line.match(this.dueRegexp);
        if (dueMatch !== null) {
            const filterDate = this.parseDate(dueMatch[2]);
            if (!filterDate.isValid()) {
                this._error = 'do not understand due date';
            }

            let filter;
            if (dueMatch[1] === 'before') {
                filter = (task: Task) =>
                    task.dueDate ? task.dueDate.isBefore(filterDate) : false;
            } else if (dueMatch[1] === 'after') {
                filter = (task: Task) =>
                    task.dueDate ? task.dueDate.isAfter(filterDate) : false;
            } else {
                filter = (task: Task) =>
                    task.dueDate ? task.dueDate.isSame(filterDate) : false;
            }

            this._filters.push(filter);
        } else {
            this._error = 'do not understand query filter (due date)';
        }
    }

    private parseDoneFilter({ line }: { line: string }): void {
        const doneMatch = line.match(this.doneRegexp);
        if (doneMatch !== null) {
            const filterDate = this.parseDate(doneMatch[2]);
            if (!filterDate.isValid()) {
                this._error = 'do not understand done date';
            }

            let filter;
            if (doneMatch[1] === 'before') {
                filter = (task: Task) =>
                    task.doneDate ? task.doneDate.isBefore(filterDate) : false;
            } else if (doneMatch[1] === 'after') {
                filter = (task: Task) =>
                    task.doneDate ? task.doneDate.isAfter(filterDate) : false;
            } else {
                filter = (task: Task) =>
                    task.doneDate ? task.doneDate.isSame(filterDate) : false;
            }

            this._filters.push(filter);
        }
    }

    private parsePathFilter({ line }: { line: string }): void {
        const pathMatch = line.match(this.pathRegexp);
        if (pathMatch !== null) {
            const filterMethod = pathMatch[1];
            if (filterMethod === 'includes') {
                this._filters.push((task: Task) =>
                    task.path.includes(pathMatch[2]),
                );
            } else if (pathMatch[1] === 'does not include') {
                this._filters.push(
                    (task: Task) => !task.path.includes(pathMatch[2]),
                );
            } else {
                this._error = 'do not understand query filter (path)';
            }
        } else {
            this._error = 'do not understand query filter (path)';
        }
    }

    private parseDescriptionFilter({ line }: { line: string }): void {
        const descriptionMatch = line.match(this.descriptionRegexp);
        if (descriptionMatch !== null) {
            const filterMethod = descriptionMatch[1];
            if (filterMethod === 'includes') {
                this._filters.push((task: Task) =>
                    this.stringIncludesCaseInsensitive(
                        task.description,
                        descriptionMatch[2],
                    ),
                );
            } else if (descriptionMatch[1] === 'does not include') {
                this._filters.push(
                    (task: Task) =>
                        !this.stringIncludesCaseInsensitive(
                            task.description,
                            descriptionMatch[2],
                        ),
                );
            } else {
                this._error = 'do not understand query filter (description)';
            }
        } else {
            this._error = 'do not understand query filter (description)';
        }
    }

    private parseHeadingFilter({ line }: { line: string }): void {
        const headingMatch = line.match(this.headingRegexp);
        if (headingMatch !== null) {
            const filterMethod = headingMatch[1].toLowerCase();
            if (filterMethod === 'includes') {
                this._filters.push(
                    (task: Task) =>
                        task.precedingHeader !== null &&
                        this.stringIncludesCaseInsensitive(
                            task.precedingHeader,
                            headingMatch[2],
                        ),
                );
            } else if (headingMatch[1] === 'does not include') {
                this._filters.push(
                    (task: Task) =>
                        task.precedingHeader !== null &&
                        !this.stringIncludesCaseInsensitive(
                            task.precedingHeader,
                            headingMatch[2],
                        ),
                );
            } else {
                this._error = 'do not understand query filter (heading)';
            }
        } else {
            this._error = 'do not understand query filter (heading)';
        }
    }

    private parseLimit({ line }: { line: string }): void {
        const limitMatch = line.match(this.limitRegexp);
        if (limitMatch !== null) {
            // limitMatch[2] is per regex always digits and therefore parsable.
            const limit = Number.parseInt(limitMatch[2], 10);
            this._limit = limit;
        } else {
            this._error = 'do not understand query limit';
        }
    }

    private parseSortBy({ line }: { line: string }): void {
        const fieldMatch = line.match(this.sortByRegexp);
        if (fieldMatch !== null) {
            this._sorting.push(fieldMatch[1] as Sorting);
        } else {
            this._error = 'do not understand query sorting';
        }
    }

    private parseDate(input: string): moment.Moment {
        // Using start of date to correctly match on comparison with other dates (like equality).
        return window.moment(chrono.parseDate(input)).startOf('day');
    }

    private stringIncludesCaseInsensitive(
        haystack: string,
        needle: string,
    ): boolean {
        return haystack
            .toLocaleLowerCase()
            .includes(needle.toLocaleLowerCase());
    }
}
