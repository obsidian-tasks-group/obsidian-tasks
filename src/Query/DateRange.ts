import type { Moment } from 'moment';

/**
 * Represent an inclusive span of time between two days at 00:00 local time.
 */
export class DateRange {
    start: Moment;
    end: Moment;

    /**
     * Builds the date range. If start is after the end, the dates will be automatically reversed.
     * Start and end will be saved at 00:00 local time.
     * The stored values of are mutable.
     * Note that there is no validation of the start and end moment. They can be checked with start.isValid() and end.isValid().
     * @param start
     * @param end
     */
    constructor(start: Moment, end: Moment) {
        this.start = start;
        this.end = end;

        if (end.isBefore(start)) {
            this.start = end;
            this.end = start;
        }

        // Dates shall be at midnight eg 00:00
        this.start = this.start.startOf('day');
        this.end = this.end.startOf('day');
    }

    /**
     * Builds a date range relative to today like this week/month/quarter/year.
     * @example <caption> construct a date range containing the current month.</caption>
     * const thisMonth = DateRange.buildRelative('month');
     * @param range one of 'week', 'month', 'quarter', 'year'
     * @returns
     */
    public static buildRelative(range: moment.unitOfTime.StartOf) {
        // Treat all weeks as ISO 8601 weeks
        const unitOfTime = range === 'week' ? 'isoWeek' : range;

        return new DateRange(
            window.moment().startOf(unitOfTime).startOf('day'),
            window.moment().endOf(unitOfTime).startOf('day'),
        );
    }

    /**
     * Builds an invalid date range with invalid momentks objects (https://momentjscom.readthedocs.io/en/latest/moment/09-utilities/02-invalid/).
     * For example if the parsing has gone wrong.
     * @returns
     */
    public static buildInvalid(): DateRange {
        return new DateRange(window.moment.invalid(), window.moment.invalid());
    }

    /**
     * Check if the both dates of the date range are valid in terms of momentjs.
     * @returns true if both dates are valid
     */
    public isValid(): boolean {
        return this.start.isValid() && this.end.isValid();
    }

    /**
     * Move the date range to the previous week/month/quarter/year.
     * Make sure that the duration in the parameters matches the range size, for example subtracting a month from a week long range was not tested.
     * @param duration one of 'week'/'month'/'quarter'/'year'
     */
    public moveToPrevious(duration: moment.unitOfTime.DurationConstructor) {
        const delta = window.moment.duration(1, duration);
        this.start.subtract(delta);
        this.end.subtract(delta);

        if (duration === 'month' || duration === 'quarter') {
            // Month and quarter durations in days may differ (28/30/31 days).
            // We will need to adjust the end.
            this.end = this.end.endOf(duration).startOf('day');
        }
    }

    /**
     * Move the date range to the next week/month/quarter/year.
     * Make sure that the duration in the parameters matches the range size, for example subtracting a month from a week long range was not tested.
     * @param duration one of 'week'/'month'/'quarter'/'year'
     */
    public moveToNext(duration: moment.unitOfTime.DurationConstructor) {
        const delta = window.moment.duration(1, duration);
        this.start.add(delta);
        this.end.add(delta);

        if (duration === 'month' || duration === 'quarter') {
            // Month and quarter durations in days may differ (28/30/31 days).
            // We will need to adjust the end.
            this.end = this.end.endOf(duration).startOf('day');
        }
    }
}
