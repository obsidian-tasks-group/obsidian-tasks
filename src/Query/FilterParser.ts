import type { Field } from './Filter/Field';
import { DescriptionField } from './Filter/DescriptionField';
import { CreatedDateField } from './Filter/CreatedDateField';
import { DoneDateField } from './Filter/DoneDateField';
import { DueDateField } from './Filter/DueDateField';
import { ExcludeSubItemsField } from './Filter/ExcludeSubItemsField';
import { HeadingField } from './Filter/HeadingField';
import { PathField } from './Filter/PathField';
import { PriorityField } from './Filter/PriorityField';
import { ScheduledDateField } from './Filter/ScheduledDateField';
import { StartDateField } from './Filter/StartDateField';
import { HappensDateField } from './Filter/HappensDateField';
import { RecurringField } from './Filter/RecurringField';
import { StatusField } from './Filter/StatusField';
import { TagsField } from './Filter/TagsField';
import { BooleanField } from './Filter/BooleanField';
import { FilenameField } from './Filter/FilenameField';
import { UrgencyField } from './Filter/UrgencyField';
import { StatusNameField } from './Filter/StatusNameField';
import { StatusTypeField } from './Filter/StatusTypeField';

import { RecurrenceField } from './Filter/RecurrenceField';
import type { FilterOrErrorMessage } from './Filter/Filter';
import type { Sorter } from './Sorter';
import type { Grouper } from './Grouper';
import { FolderField } from './Filter/FolderField';
import { RootField } from './Filter/RootField';
import { BacklinkField } from './Filter/BacklinkField';

// When parsing a query the fields are tested one by one according to this order.
// Since BooleanField is a meta-field, which needs to aggregate a few fields together, it is intended to
// be kept last.
// When adding new fields keep this order in mind, putting fields that are more specific before fields that
// may contain them, and keep BooleanField last.
const fieldCreators: EndsWith<BooleanField> = [
    () => new StatusNameField(), // status.name is before status, to avoid ambiguity
    () => new StatusTypeField(), // status.type is before status, to avoid ambiguity
    () => new StatusField(),
    () => new RecurringField(),
    () => new PriorityField(),
    () => new HappensDateField(),
    () => new CreatedDateField(),
    () => new StartDateField(),
    () => new ScheduledDateField(),
    () => new DueDateField(),
    () => new DoneDateField(),
    () => new PathField(),
    () => new FolderField(),
    () => new RootField(),
    () => new BacklinkField(),
    () => new DescriptionField(),
    () => new TagsField(),
    () => new HeadingField(),
    () => new ExcludeSubItemsField(),
    () => new FilenameField(),
    () => new UrgencyField(),
    () => new RecurrenceField(),
    () => new BooleanField(), // --- Please make sure to keep BooleanField last (see comment above) ---
];

// This type helps verify that BooleanField is kept last
type EndsWith<End, T extends Field = Field> = [...Array<() => T>, () => End];

export function parseFilter(filterString: string): FilterOrErrorMessage | null {
    for (const creator of fieldCreators) {
        const field = creator();
        if (field.canCreateFilterForLine(filterString)) return field.createFilterOrErrorMessage(filterString);
    }
    return null;
}

export function parseSorter(sorterString: string): Sorter | null {
    // New style parsing, using sorting which is done by the Field classes.

    // Optimisation: Check whether line begins with 'sort by'
    const sortByRegexp = /^sort by /;
    if (sorterString.match(sortByRegexp) === null) {
        return null;
    }

    // See if any of the fields can parse the line.
    for (const creator of fieldCreators) {
        const field = creator();
        const sorter = field.createSorterFromLine(sorterString);
        if (sorter) {
            return sorter;
        }
    }
    return null;
}

export function parseGrouper(line: string): Grouper | null {
    // New style parsing, using grouping which is done by the Field classes.

    // Optimisation: Check whether line begins with 'group by'
    const groupByRegexp = /^group by /;
    if (line.match(groupByRegexp) === null) {
        return null;
    }

    // See if any of the fields can parse the line.
    for (const creator of fieldCreators) {
        const field = creator();
        const grouper = field.createGrouperFromLine(line);
        if (grouper) {
            return grouper;
        }
    }
    return null;
}
