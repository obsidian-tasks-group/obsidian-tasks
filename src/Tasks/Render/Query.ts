import moment from 'moment';
import { Status } from 'Tasks/Status';
import { DATE_FORMAT, Task } from '../Task';

export class Query {
    public readonly filters: ((task: Task) => boolean)[];
    private readonly noDueString = 'no due date';
    private readonly dueRegexp = /due (before|after|on) (\d{4}-\d{2}-\d{2})/;
    private readonly doneString = 'done';
    private readonly notDoneString = 'not done';
    private readonly doneRegexp = /done (before|after|on) (\d{4}-\d{2}-\d{2})/;
    private readonly pathRegexp = /path (includes|does not include) (.*)/;
    private readonly excludeSubItemsString = 'exclude sub-items';

    constructor({ source }: { source: string }) {
        this.filters = this.parseFilters({ source });
    }

    private parseFilters({
        source,
    }: {
        source: string;
    }): ((task: Task) => boolean)[] {
        const filters: ((task: Task) => boolean)[] = [];
        const sourceLines = source.split('\n').map(line => line.trim());

        for (const sourceLine of sourceLines) {
            if (sourceLine === '') {
                continue;
            }

            if (sourceLine === this.doneString) {
                filters.push((task) => task.status === Status.DONE);
            } else if (sourceLine === this.notDoneString) {
                filters.push((task) => task.status !== Status.DONE);
            } else if (sourceLine === this.noDueString) {
                filters.push((task) => task.dueDate === undefined);
            } else if (this.dueRegexp.test(sourceLine)) {
                const filter = this.parseDueFilter({ line: sourceLine });
                if (filter !== undefined) {
                    filters.push(filter);
                }
            } else if (this.doneRegexp.test(sourceLine)) {
                const filter = this.parseDoneFilter({ line: sourceLine });
                if (filter !== undefined) {
                    filters.push(filter);
                }
            } else if (this.pathRegexp.test(sourceLine)) {
                const filter = this.parsePathFilter({ line: sourceLine });
                if (filter !== undefined) {
                    filters.push(filter);
                }
            } else if (sourceLine === this.excludeSubItemsString) {
                filters.push((task) => task.indentation === '');
            } else {
                console.error('Tasks: do not understand filter: ' + sourceLine);
            }
        }

        return filters;
    }

    private parseDueFilter({
        line,
    }: {
        line: string;
    }): ((task: Task) => boolean) | undefined {
        const dueMatch = line.match(this.dueRegexp);
        if (dueMatch !== null) {
            let filter;
            const filterDate = moment(dueMatch[2], DATE_FORMAT);
            if (dueMatch[1] === 'before') {
                filter = (task: Task) =>
                    task.dueDate ? task.dueDate.isBefore(filterDate) : false;
            } else if (dueMatch[1] === 'after') {
                filter = (task: Task) =>
                    task.dueDate ? task.dueDate.isAfter(filterDate) : false;
            } else {
                filter = (task: Task) =>
                    task.dueDate ? task.dueDate.isSame(filterDate) : false;
            }

            return filter;
        }

        return undefined;
    }

    private parseDoneFilter({
        line,
    }: {
        line: string;
    }): ((task: Task) => boolean) | undefined {
        const doneMatch = line.match(this.doneRegexp);
        if (doneMatch !== null) {
            let filter;
            const filterDate = moment(doneMatch[2], DATE_FORMAT);
            if (doneMatch[1] === 'before') {
                filter = (task: Task) =>
                    task.doneDate ? task.doneDate.isBefore(filterDate) : false;
            } else if (doneMatch[1] === 'after') {
                filter = (task: Task) =>
                    task.doneDate ? task.doneDate.isAfter(filterDate) : false;
            } else {
                filter = (task: Task) =>
                    task.doneDate ? task.doneDate.isSame(filterDate) : false;
            }

            return filter;
        }

        return undefined;
    }

    private parsePathFilter({
        line,
    }: {
        line: string;
    }): ((task: Task) => boolean) | undefined {
        const pathMatch = line.match(this.pathRegexp);
        if (pathMatch !== null) {
            let filter;
            const filterMethod = pathMatch[1];
            if (filterMethod === 'includes') {
                filter = (task: Task) =>
                    task.path.includes(pathMatch[2]);
            } else if (pathMatch[1] === 'does not include') {
                filter = (task: Task) =>
                    !task.path.includes(pathMatch[2]);
            } else {
                console.error('Tasks: do not understand path query: ' + line);
                return undefined;
            }

            return filter;
        }

        return undefined;
    }
}
