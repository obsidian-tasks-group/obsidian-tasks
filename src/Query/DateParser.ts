import * as chrono from 'chrono-node';
import type { Moment } from 'moment';
import { DateRange } from './DateRange';

export class DateParser {
    public static parseDate(input: string, forwardDate: boolean = false): Moment {
        // Using start of day to correctly match on comparison with other dates (like equality).
        return window
            .moment(
                chrono.parseDate(input, undefined, {
                    forwardDate: forwardDate,
                }),
            )
            .startOf('day');
    }

    /**
     * Parse a line and extract a pair of dates, returned in a tuple, sorted by date.
     * @param input - any pair of dates, separate by one or more spaces '17 August 2013 19 August 2013',
     *                or a single date.
     * @param forwardDate - if true, and date is ambiguous, chrono will return dates in the future
     * @return - A Tuple of dates. If both input dates are invalid, then both output dates will be invalid.
     */
    public static parseDateRange(input: string, forwardDate: boolean = false): DateRange {
        const dateRangeParsers = [
            // Try parsing a relative date range like 'current month'
            DateParser.parseRelativeDateRange,
            // Try '2022-W10' otherwise
            DateParser.parseNumberedDateRange,
            // If previous failed, fallback on absolute date range with chrono
            DateParser.parseAbsoluteDateRange,
        ];

        for (const parser of dateRangeParsers) {
            const parsedDateRange = parser(input, forwardDate);
            if (parsedDateRange.isValid()) {
                return parsedDateRange;
            }
        }

        // If nothing worked return and invalid date range
        return DateRange.buildInvalid();
    }

    private static parseAbsoluteDateRange(input: string, forwardDate: boolean): DateRange {
        const result = chrono.parse(input, undefined, {
            forwardDate: forwardDate,
        });

        // Check chrono parsing
        if (result.length === 0) {
            return DateRange.buildInvalid();
        }

        const startDate = result[0].start;
        const endDate = result[1] && result[1].start ? result[1].start : startDate;
        const start = window.moment(startDate.date());
        const end = window.moment(endDate.date());

        return new DateRange(start, end);
    }

    private static parseRelativeDateRange(input: string, _forwardDate: boolean): DateRange {
        const relativeDateRangeRegexp = /(last|this|next) (week|month|quarter|year)/;
        const relativeDateRangeMatch = input.match(relativeDateRangeRegexp);
        if (relativeDateRangeMatch && relativeDateRangeMatch.length === 3) {
            const lastThisNext = relativeDateRangeMatch[1];
            const range = relativeDateRangeMatch[2];

            const dateRange = DateRange.buildRelative(range as moment.unitOfTime.StartOf);

            switch (lastThisNext) {
                case 'last':
                    dateRange.moveToPrevious(range as moment.unitOfTime.DurationConstructor);
                    break;
                case 'next':
                    dateRange.moveToNext(range as moment.unitOfTime.DurationConstructor);
                    break;
            }

            return dateRange;
        }

        return DateRange.buildInvalid();
    }

    private static parseNumberedDateRange(input: string, _forwardDate: boolean): DateRange {
        const parsingVectors: [RegExp, string, moment.unitOfTime.StartOf][] = [
            [/^\s*[0-9]{4}\s*$/, 'YYYY', 'year'],
            [/^\s*[0-9]{4}-Q[1-4]\s*$/, 'YYYY-Q', 'quarter'],
            [/^\s*[0-9]{4}-[0-9]{2}\s*$/, 'YYYY-MM', 'month'],
            [/^\s*[0-9]{4}-W[0-9]{2}\s*$/, 'YYYY-WW', 'isoWeek'],
        ];

        for (const [regexp, dateFormat, range] of parsingVectors) {
            const matched = input.match(regexp);
            if (matched) {
                // RegExps allow spaces (\s*), remove them before calling window.moment()
                const date = matched[0].trim();
                return new DateRange(
                    window.moment(date, dateFormat).startOf(range),
                    window.moment(date, dateFormat).endOf(range),
                );
            }
        }

        return DateRange.buildInvalid();
    }
}
