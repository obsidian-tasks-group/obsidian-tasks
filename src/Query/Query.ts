import { LayoutOptions } from '../TaskLayout';
import type { Task } from '../Task';
import type { IQuery } from '../IQuery';
import { getSettings } from '../Config/Settings';
import { Sort } from './Sort';
import type { Sorter } from './Sorter';
import { TaskGroups } from './TaskGroups';
import * as FilterParser from './FilterParser';
import type { Grouper } from './Grouper';
import type { Filter } from './Filter/Filter';

export class Query implements IQuery {
    public source: string;

    private _limit: number | undefined = undefined;
    private _taskGroupLimit: number | undefined = undefined;
    private _layoutOptions: LayoutOptions = new LayoutOptions();
    private _filters: Filter[] = [];
    private _error: string | undefined = undefined;
    private _sorting: Sorter[] = [];
    private _grouping: Grouper[] = [];

    private readonly hideOptionsRegexp =
        /^(hide|show) (task count|backlink|priority|created date|start date|scheduled date|done date|due date|recurrence rule|edit button|urgency)/;
    private readonly shortModeRegexp = /^short/;
    private readonly explainQueryRegexp = /^explain/;

    private readonly limitRegexp = /^limit (groups )?(to )?(\d+)( tasks?)?/;

    private readonly commentRegexp = /^#.*/;

    constructor({ source }: { source: string }) {
        this.source = source;
        source
            .split('\n')
            .map((line: string) => line.trim())
            .forEach((line: string) => {
                switch (true) {
                    case line === '':
                        break;
                    case this.shortModeRegexp.test(line):
                        this._layoutOptions.shortMode = true;
                        break;
                    case this.explainQueryRegexp.test(line):
                        this._layoutOptions.explainQuery = true;
                        break;
                    case this.limitRegexp.test(line):
                        this.parseLimit({ line });
                        break;
                    case this.parseSortBy({ line }):
                        break;
                    case this.parseGroupBy({ line }):
                        break;
                    case this.hideOptionsRegexp.test(line):
                        this.parseHideOptions({ line });
                        break;
                    case this.commentRegexp.test(line):
                        // Comment lines are ignored
                        break;
                    case this.parseFilter(line):
                        break;
                    default:
                        this._error = `do not understand query: ${line}`;
                }
            });
    }

    /**
     *
     * Appends {@link q2} to this query.
     *
     * @note At time of writing, this query language appears to play nicely with combining queries.
     *
     * More formally, the concatenation operation on the query language:
     *     * Is closed (concatenating two queries is another valid query)
     *     * Is not commutative (q1.append(q2) !== q2.append(q1))
     *
     * And the semantics of the combination are straight forward:
     *     * Combining two queries appends their filters
     *           (assuming that the filters are pure functions, filter concatenation is commutative)
     *     * Combining two queries appends their sorting instructions. (this is not commutative)
     *     * Combining two queries appends their grouping instructions. (this is not commutative)
     *     * Successive limit instructions overwrite previous ones.
     *
     * @param {Query} q2
     * @return {Query} The combined query
     */
    public append(q2: Query): Query {
        if (this.source === '') return q2;
        if (q2.source === '') return this;
        return new Query({ source: `${this.source}\n${q2.source}` });
    }

    /**
     * Generate a text description of the contents of this query.
     *
     * This does not show any global filter and global query.
     * Use {@link explainResults} if you want to see any global query and global filter as well.
     */
    public explainQuery(): string {
        let result = '';

        const numberOfFilters = this.filters.length;
        if (numberOfFilters === 0) {
            result += 'No filters supplied. All tasks will match the query.';
        } else {
            for (let i = 0; i < numberOfFilters; i++) {
                if (i > 0) result += '\n';
                result += this.filters[i].explainFilterIndented('');
            }
        }
        result += this.explainQueryLimits();

        const { debugSettings } = getSettings();
        if (debugSettings.ignoreSortInstructions) {
            result +=
                "\n\nNOTE: All sort instructions, including default sort order, are disabled, due to 'ignoreSortInstructions' setting.";
        }

        return result;
    }

    private explainQueryLimits() {
        let result = '';

        function getPluralisedText(limit: number) {
            let text = `\n\nAt most ${limit} task`;
            if (limit !== 1) {
                text += 's';
            }
            return text;
        }

        if (this._limit !== undefined) {
            result += getPluralisedText(this._limit);
            result += '.\n';
        }

        if (this._taskGroupLimit !== undefined) {
            result += getPluralisedText(this._taskGroupLimit);
            result += ' per group (if any "group by" options are supplied).\n';
        }
        return result;
    }

    public get limit(): number | undefined {
        return this._limit;
    }

    public get layoutOptions(): LayoutOptions {
        return this._layoutOptions;
    }

    public get filters(): Filter[] {
        return this._filters;
    }

    public get sorting() {
        return this._sorting;
    }

    /**
     * Return the {@link Grouper} objects that represent any `group by` instructions in the tasks block.
     */
    public get grouping(): Grouper[] {
        return this._grouping;
    }

    public get error(): string | undefined {
        return this._error;
    }

    public applyQueryToTasks(tasks: Task[]): TaskGroups {
        this.filters.forEach((filter) => {
            tasks = tasks.filter(filter.filterFunction);
        });

        const { debugSettings } = getSettings();
        const tasksSorted = debugSettings.ignoreSortInstructions ? tasks : Sort.by(this.sorting, tasks);
        const tasksSortedLimited = tasksSorted.slice(0, this.limit);

        const taskGroups = new TaskGroups(this.grouping, tasksSortedLimited);

        if (this._taskGroupLimit !== undefined) {
            taskGroups.applyTaskLimit(this._taskGroupLimit);
        }

        return taskGroups;
    }

    private parseHideOptions({ line }: { line: string }): void {
        const hideOptionsMatch = line.match(this.hideOptionsRegexp);
        if (hideOptionsMatch !== null) {
            const hide = hideOptionsMatch[1] === 'hide';
            const option = hideOptionsMatch[2];

            switch (option) {
                case 'task count':
                    this._layoutOptions.hideTaskCount = hide;
                    break;
                case 'backlink':
                    this._layoutOptions.hideBacklinks = hide;
                    break;
                case 'priority':
                    this._layoutOptions.hidePriority = hide;
                    break;
                case 'created date':
                    this._layoutOptions.hideCreatedDate = hide;
                    break;
                case 'start date':
                    this._layoutOptions.hideStartDate = hide;
                    break;
                case 'scheduled date':
                    this._layoutOptions.hideScheduledDate = hide;
                    break;
                case 'due date':
                    this._layoutOptions.hideDueDate = hide;
                    break;
                case 'done date':
                    this._layoutOptions.hideDoneDate = hide;
                    break;
                case 'recurrence rule':
                    this._layoutOptions.hideRecurrenceRule = hide;
                    break;
                case 'edit button':
                    this._layoutOptions.hideEditButton = hide;
                    break;
                case 'urgency':
                    this._layoutOptions.hideUrgency = hide;
                    break;
                default:
                    this._error = 'do not understand hide/show option';
            }
        }
    }

    private parseFilter(line: string) {
        const filterOrError = FilterParser.parseFilter(line);
        if (filterOrError != null) {
            if (filterOrError.filter) this._filters.push(filterOrError.filter);
            else this._error = filterOrError.error;
            return true;
        }
        return false;
    }

    private parseLimit({ line }: { line: string }): void {
        const limitMatch = line.match(this.limitRegexp);
        if (limitMatch === null) {
            this._error = 'do not understand query limit';
            return;
        }

        // limitMatch[3] is per regex always digits and therefore parsable.
        const limitFromLine = Number.parseInt(limitMatch[3], 10);

        if (limitMatch[1] !== undefined) {
            this._taskGroupLimit = limitFromLine;
        } else {
            this._limit = limitFromLine;
        }
    }

    private parseSortBy({ line }: { line: string }): boolean {
        const sortingMaybe = FilterParser.parseSorter(line);
        if (sortingMaybe) {
            this._sorting.push(sortingMaybe);
            return true;
        }
        return false;
    }

    /**
     * Parsing of `group by` lines, for grouping that is implemented in the {@link Field}
     * classes.
     *
     * @param line
     * @private
     */
    private parseGroupBy({ line }: { line: string }): boolean {
        const groupingMaybe = FilterParser.parseGrouper(line);
        if (groupingMaybe) {
            this._grouping.push(groupingMaybe);
            return true;
        }
        return false;
    }
}
