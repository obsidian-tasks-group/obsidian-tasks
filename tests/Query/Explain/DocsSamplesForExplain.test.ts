/**
 * @jest-environment jsdom
 */
import { verify } from 'approvals/lib/Providers/Jest/JestApprovals';
import moment from 'moment';
import { Query } from '../../../src/Query/Query';

window.moment = moment;

describe('explain', () => {
    beforeAll(() => {
        jest.useFakeTimers();
        jest.setSystemTime(new Date(2022, 9, 21)); // 2022-10-21
    });

    afterAll(() => {
        jest.useRealTimers();
    });

    it('expands dates', () => {
        // Arrange
        const instructions: string = `
            "
            scheduled after 2 years ago
            due before tomorrow
            explain
            "
            `;
        const query = new Query({ source: instructions });

        // Act
        const text = query.explainQuery();

        // Assert
        verify(text);
    });
});
