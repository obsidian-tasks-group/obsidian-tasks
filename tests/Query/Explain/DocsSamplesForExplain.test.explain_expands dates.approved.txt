All of:
  scheduled after 2 years ago =>
    scheduled date is after 2020-10-21 (Wednesday 21st October 2020)

  due before tomorrow =>
    due date is before 2022-10-22 (Saturday 22nd October 2022)
